### PR TITLE
Set status 503 when backup is in progress. Fixes #5221

### LINKED ISSUE #5221: "Go backup in progress"-page saved in artifact that was fetched during GoCD-backup
##### Issue Type

- Bug report

##### Summary

We encountered an issue where a task to fetch an artifact (a configuration file) from a previous stage was run at the same time as our GoCD server was backing up. The fetched file was still created, but instead of containing the expected configuration it contained the "Go backup in progress."-html page.

However, GoCD still ran our deploy step after this and we only noticed that something was wrong a while later when trying to load the deployed application.

I found this in our build log:
```plain
[go] Task: fetch artifact [src/config.js] => [src] from [app/build/build]took: 0.19s
[go] Fetching artifact [src/config.js] from [app/1572/build/latest/build]
[go] [WARN] The md5checksum value of the artifact [src/config.js] was not found on the server. Hence, Go could not verify the integrity of its contents.
[go] Saved artifact to [pipelines/app/src/config.js] without verifying the integrity of its contents.
[go] Task status: passed, took: 0.19s
```

I assume that the server backing up also resulted in the md5 checksum not being found. I read in the documentation that this is the intended behavior when a checksum could not be found, but it would be nice with an option to require checksums and fail the build otherwise. I was gonna submit that as a feature proposal, but the corrupted artifact seems odd enough that a bug report seems justified.

###### Basic environment details

* Go Version: `18.8.0 (7433-4bf750d09bf4dba76a2b7ce0d72e88ecdfdbd96a)`
* JAVA Version: `1.8.0_152`
* OS: `Linux 4.14.62-65.117.amzn1.x86_64`

### PR DESCRIPTION
**Pull request by ketan**
Related to #5221

### CODE REVIEW

**Comment by ketan** [comment]
Initial PR, I feel we should consider sending a 201 accepted status for requests to files, so that agents continue to "wait" and builds do not fail.

WDYT?

**Comment by arvindsv** [comment]
Unless they wait indefinitely, it might not be enough. As far as I know, they try 3 times and then fail anyway.

**Comment by ketan** [comment]
Alternatively, let calls to `/go/files` pass-through, since they do not affect the DB in any way.

File: server/src/main/java/com/thoughtworks/go/server/web/BackupFilter.java [hunk 1]
```diff
@@ -68,6 +68,12 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         }
 
         String url = ((HttpServletRequest) request).getRequestURI();
+
+        if (isFilesAPI(url)) {
```
Comment (arvindsv) [inline server/src/main/java/com/thoughtworks/go/server/web/BackupFilter.java]: @ketan Rethinking this, I'm not sure if this is a good idea. Can't calls [such as this](https://github.com/gocd/gocd/blob/aa4c0d44fd65b2a7c5c694f923f202cb6f4ecc7c/server/src/main/java/com/thoughtworks/go/server/newsecurity/providers/AbstractPluginAuthenticationProvider.java#L121) affect the DB, while a backup is in progress?

Comment (ketan) [reply server/src/main/java/com/thoughtworks/go/server/web/BackupFilter.java]: Fair point, I'll yank that commit out.

Comment (ketan) [reply server/src/main/java/com/thoughtworks/go/server/web/BackupFilter.java]: Yanked. Can someone take a look?

**Comment by arvindsv** [comment]
Looks ok to me. Tried with a config such as this:

```xml
  <pipelines group="defaultGroup">
    <pipeline name="Upstream">
      <materials>
        <git url="https://github.com/arvindsv/faketime.git" />
      </materials>
      <stage name="stage1">
        <jobs>
          <job name="defaultJob">
            <tasks>
              <exec command="touch">
                <arg>abc.txt</arg>
                <runif status="passed" />
              </exec>
            </tasks>
            <artifacts>
              <artifact type="build" src="abc.txt" />
            </artifacts>
          </job>
        </jobs>
      </stage>
    </pipeline>
    <pipeline name="Downstream">
      <materials>
        <pipeline pipelineName="Upstream" stageName="stage1" />
      </materials>
      <stage name="defaultStage">
        <jobs>
          <job name="defaultJob">
            <tasks>
              <exec command="bash">
                <arg>-c</arg>
                <arg>while [ ! -f "/tmp/a" ]; do echo "Waiting ..."; sleep 1; done</arg>
                <runif status="passed" />
              </exec>
              <fetchartifact artifactOrigin="gocd" srcfile="abc.txt" pipeline="Upstream" stage="stage1" job="defaultJob">
                <runif status="passed" />
              </fetchartifact>
            </tasks>
          </job>
        </jobs>
      </stage>
    </pipeline>
  </pipelines>
```

... and suspended the thread at a breakpoint in [BackupService#doPerformBackup](https://github.com/gocd/gocd/blob/aa4c0d44fd65b2a7c5c694f923f202cb6f4ecc7c/server/src/main/java/com/thoughtworks/go/server/service/BackupService.java#L128).

Saw fetch artifact fail like this:

```
WARN  [scheduler-1] DownloadAction:70 - Could not fetch artifact https://127.0.0.1:8154/go/remoting/files/Upstream/1/stage1/1/defaultJob/cruise-output/md5.checksum. Pausing 14 seconds to retry. Error was : Unsuccessful response '503' from the server
WARN  [scheduler-1] DownloadAction:70 - Could not fetch artifact https://127.0.0.1:8154/go/remoting/files/Upstream/1/stage1/1/defaultJob/cruise-output/md5.checksum. Pausing 25 seconds to retry. Error was : Unsuccessful response '503' from the server
WARN  [scheduler-1] DownloadAction:70 - Could not fetch artifact https://127.0.0.1:8154/go/remoting/files/Upstream/1/stage1/1/defaultJob/cruise-output/md5.checksum. Pausing 37 seconds to retry. Error was : Unsuccessful response '503' from the server
ERROR [scheduler-1] DownloadAction:58 - Giving up fetching resource 'https://127.0.0.1:8154/go/remoting/files/Upstream/1/stage1/1/defaultJob/cruise-output/md5.checksum'. Tried 4 times and failed.
Giving up fetching resource 'https://127.0.0.1:8154/go/remoting/files/Upstream/1/stage1/1/defaultJob/cruise-output/md5.checksum'. Tried 4 times and failed.
  at com.thoughtworks.go.domain.DownloadAction.perform(DownloadAction.java:59)
  at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.pullArtifact(FetchArtifactBuilder.java:67)
  at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.downloadChecksumFile(FetchArtifactBuilder.java:63)
  at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.fetch(FetchArtifactBuilder.java:52)
  at com.thoughtworks.go.publishers.GoArtifactsManipulator.fetch(GoArtifactsManipulator.java:172)
  at com.thoughtworks.go.work.DefaultGoPublisher.fetch(DefaultGoPublisher.java:76)
  at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.build(FetchArtifactBuilder.java:48)
  at com.thoughtworks.go.remote.work.Builders.build(Builders.java:74)
  at com.thoughtworks.go.remote.work.BuildWork.execute(BuildWork.java:220)
  at com.thoughtworks.go.remote.work.BuildWork.buildJob(BuildWork.java:191)
  at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:137)
  at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:82)
  at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:54)
  at com.thoughtworks.go.agent.AgentHTTPClientController.retrieveWork(AgentHTTPClientController.java:134)
  at com.thoughtworks.go.agent.AgentHTTPClientController.work(AgentHTTPClientController.java:109)
  at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:85)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Then, once the backup finished, the job rescheduled and finished.

Would have been nice to have had a way to test this by slowing down the backup. Currently, it can only be tested by suspending the thread.

**Comment by arvindsv** [comment]
Also ran a few `curl` commands and they all returned 503.

**Comment by arvindsv** [comment]
@ketan One thing I was thinking of. It would have been nice to see "Service Unavailable. Backup in progress" in the agent logs.

**Pull request merged**